### ISSUE TITLE
KanesMethod solves kinematics explicitly even if not asked

### ISSUE THREAD

**Issue by zouhairm** [sympy/sympy#22626]
While dynamic equations are left in explicit form `M [u].diff(t) = f(q, u)`, the kinematic equations are solved explicitly [here](https://github.com/sympy/sympy/blob/0e4bab73562d6f90fcdf5fa079731f26a455347c/sympy/physics/mechanics/kane.py#L259):
```
            f_k = k_kqdot.LUsolve(f_k)
            k_ku = k_kqdot.LUsolve(k_ku)
            k_kqdot = eye(len(qdot))
```

The end result is that in `mass_matrix_full` the columns associated with kinematics are just the identity. This works fine most of the time, but can be unnecessarily slow if an explicit form of the equations is not desired, not to mention that it requires simplification of the explicit right hand side.

Would it make sense to avoid this step unless explicitly requested by the initialization function?

Happy to submit a PR if others here suggest how to best control that behavior.

**Comment by moorepants** [comment]
If `u=q'` is chosen then it isn't necessary to solve the equations for q' because it is trivial. So, yes we could short circuit that `LUsolve()`. I think we didn't because even for very large number of generalized coordinates, that LUsolve should happen instantaneously. I'm not sure why you say "but can be unnecessarily slow". That signals to me that you don't have a trivial definition of the generalized speeds, because that would be the only reason that LUsolve is slow.

**Comment by oscarbenjamin** [comment]
If there are cases where `LUsolve` with the identity matrix is slow then we should probably make improvements to `LUsolve` to speed that up.

**Comment by zouhairm** [comment]
Indeed I don't have a trivial definition of generalized speeds (i.e. u != q', at least not for all of the coordinates, which is common in aerospace with the generalized speeds being defined in body frame but the generalized coordinate being Euler angles)

LUsolve ends up slowing things a bit but not significantly, where I'm finding issues is in the right hand side (i.e. the kinematics part of forcing_full) since the symbolic LUSolve results in overly complicated expressions as the mass matrix is inverted. Those complex expressions cause two issues:
1. numerical issues when evaluated as there are typically trigonometric expressions with 1/cos terms that blow up
2. further complicated linearized equations 

The way I've addressed these issues so far is by using symbolic simplification on the right hand side but that is indeed the much slower part.

In contrast, if the kinematics were left in explicit form, then the slow simplification step is not needed and the expressions don't have numerical issues and linearization won't have to deal complicated expressions. LUSolve can always be done numerically when/if needed to compute the q'.

**Comment by zouhairm** [comment]
To illustrate this, here is what the equations from LUSolve look like (first 3 rows of full mass matrix and forcing function are 3 of my kinematic equations) vs. what it looks like without the LUSolve.

![image](https://user-images.githubusercontent.com/430175/145310169-4c4cedd2-6763-4ab1-8fda-f74fb52d340c.png)

Simplifying forcing_full in my case can be very long, and so I have resorted to just simplifying the kinematic portion of it and that has been an ok workaround

![image](https://user-images.githubusercontent.com/430175/145310577-47a4199b-2fe5-4f61-9e0b-370f85d9793a.png)

But I've realized that it's simpler to just keep the kinematics in their implicit form since the equations are already geared towards that, which is why I opened the issue to ask if it's something others would find useful :)

**Comment by moorepants** [comment]
There is room for improvement in KanesMethod to provide the equations (kinematics + dynamic) in a fully implicit form and skipping the kinematic equation solve internal to KanesMethod. If you want to open a PR with some ideas of how to do this, I'm happy to review. There will be a devil in the details, I'm sure.

For non-toy problems I never use any simplification routines, except for common subexpression elimination. They are always going to be slow and are only really useful if you want to read/print equations. If you want to read equations, then you are stuck with slow computation.

> numerical issues when evaluated as there are typically trigonometric expressions with 1/cos or 1/sin terms that can blow up (but that otherwise would cancel out if simplified [see below])

This is something we need to make improvements on. There are open issues about this in SymPy and PyDy. I would love to see this remedied.

**Comment by zouhairm** [comment]
@moorepants  - just submitted a PR. LMK about thoughts of whether this looks reasonable.

I'd also appreciate any guidance on how to properly test this (i.e. what functions you think need to be checked that they still behave as expected even with the flag set to keep things in implicit form)

**Comment by moorepants** [comment]
The PR is https://github.com/sympy/sympy/pull/22818

**Comment by moorepants** [comment]
I've been reading the code and trying to understand why this explicit solve of the kinematical differential equations occurs. I'm not yet sure of the answer, but, at this point, I don't think that solve should be happening. The default should be to provide the dynamical and kinematical equations in implicit forms: Ax'=b. Solving for q' and then using those expressions in the further formulation of Mu'=F will cause unnecessary increase in complexity (and possibly introduce the 1/trigfunc type of expressions). The correct solution is to probably purge this behavior. I think the linearization routines may rely on this form though, as I'm getting some breaking tests when I don't do the solve.

**Comment by moorepants** [comment]
If we remove the solves, the only relevant failing test I'm getting is this one:

```
        # This code tests our output vs. benchmark values. When r=g=m=1, the
        # critical speed (where all eigenvalues of the linearized equations are 0)
        # is 1 / sqrt(3) for the upright case.
        A = KM.linearize(A_and_B=True)[0]
        A_upright = A.subs({r: 1, g: 1, m: 1}).subs({q1: 0, q2: 0, q3: 0, u1: 0, u3: 0})
        import sympy
>       assert sympy.sympify(A_upright.subs({u2: 1 / sqrt(3)})).eigenvals() == {S.Zero: 6}
E       AssertionError: assert {-sqrt(5)/4 -... 5/8): 1, ...} == {0: 6}
E         Differing items:
E         {0: 2} != {0: 6}
E         Left contains 4 more items:
E         {-1/4 + sqrt(5)/4 - I*sqrt(sqrt(5)/8 + 5/8): 1,
E          -sqrt(5)/4 - 1/4 - I*sqrt(5/8 - sqrt(5)/8): 1,
E          -sqrt(5)/4 - 1/4 + I*sqrt(5/8 - sqrt(5)/8): 1,
E          -1/4 + sqrt(5)/4 + I*sqrt(sqrt(5)/8 + 5/8): 1}...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show

sympy/physics/mechanics/tests/test_kane.py:175: AssertionError
```

A_upright ends up with nans and zoos.

**Comment by zouhairm** [comment]
> Solving for q' and then using those expressions in the further formulation of Mu'=F will cause unnecessary increase in complexity 

Do you mean just the `f_k = k_kqdot.LUsolve(f_k)` lines, or the `self._qdot_u_map = solve_linear_system_LU`, because the latter seems to be used in a few places (I think that's needed in case there are points with velocities that have not been defined solely in terms of generalized speeds and require some kinematic terms).

> I don't think that solve should be happening. [...] The correct solution is to probably purge this behavior. 

I'm in favor of purging that behavior and it'd simplify this PR by not having an additional behavior that needs to be tested differently. I'm not familiar enough with the rolling disk problem to understand why that would fail and how to fix it, and yeah I think the linearization is indeed dependent on this.

would moving this explicit form to the linearization block make sense?
Is that something you can investigate @moorepants, or point me to the rough region in the code I should be looking at? Thanks!

**Comment by moorepants** [comment]
I think your approach to have an option is probably the smart thing to do considering backwards incompatibilities (based on us changing the expected output).

**Comment by zouhairm** [comment]
I'm still a bit unclear about what you mean by not doing the solve,
especially after your other PR. Are you suggesting that the qdot_u_map
would not be there anymore, or just the change that I made where qdot is
solved for but the matrices are not explicitly inverted and stored as such?

I can see not solving for qdot at all, but that'd be a bigger change since
it'd no longer be possible to get separated kinematics from dynamics
equations. But that'd be a bigger change than what I was thinking about

**Comment by moorepants** [comment]
> I'm still a bit unclear about what you mean by not doing the solve, especially after your other PR. Are you suggesting that the qdot_u_map would not be there anymore, or just the change that I made where qdot is solved for but the matrices are not explicitly inverted and stored as such?

I'm referring to these lines:

```python
            f_k = k_kqdot.LUsolve(f_k)
            k_ku = k_kqdot.LUsolve(k_ku)
            k_kqdot = eye(len(qdot))
```

not:

```python
            self._qdot_u_map = solve_linear_system_LU(
                    Matrix([k_kqdot.T, -(k_ku * u + f_k).T]).T, qdot)
```

I'm not yet sure what the purpose of the first set of lines is.

There are these equations:

1) kinematic: N(q, t)*q'(t) + g(u, q, t) = 0
2) dynamic: M(q, t)*u'(t) + f(u, q, t) = 0
3) constraints: h(u, q, t) = 0

To eliminate q' from equation 2), you have to solve 1) for q'. But you can probably set things up like:

1) kinematic: N(q, t)*q'(t) + g(u, q, t) = 0
2) dynamic: M(q, t)*u'(t) + f(u, q', q, t) = 0

and not have to explicitly solve (symbolically) for q'.

The easiest thing to do now, is to probably just expose N(q, t) to the user, as you have proposed. That can be helpful.

**Comment by zouhairm** [comment]
I've updated the PR to be based on your recent one.
The issue with the rolling disk was that the equations needed to be simplified before the subs of numerical values.

---

I think there is still a design issue I'm seeing in this:
`To eliminate q' from equation 2), you have to solve 1) for q'. `

My intention initially was to just get rid of those first 3 lines that forced kinematics to be in explicit form. But it looks like that if I do that, then the _qdot_u_map is no longer available, and one can't replace q' that appear in (2) with u.

The way I have the PR implemented is that the solve to get qdot_u_map still happens. This means that my kinematic equations are now explicit, but indeed looking at the dynamic ones, that solve is still generating unpleasant symbolic expressions to effectively provide dynamics that are *decoupled* from kinematics

I thought at first that just keeping the solve for qdot_u_map  can still get around the issue without having to refactor all of the code and leaving the dynamic equations to be coupled with kinematics. In my problem this seemed to work at first, but it's because I was only inspecting the kinematic matrices, now that I'm looking at the dynamic equations, indeed qdot_u_map being substituted in the equations means that the explicit kinematics are being substituted in there and I end up with complicated (and slow to manipulate) expressions.

So I think to do this properly I have to remove the use of qdot_u_map when `explicit_kinematics=False`, which requires digging into the code a bit more and the changes are not as trivial. I'll see if I can figure it out and might ask for some guidance here.

**Comment by zouhairm** [comment]
Ah ! I think I was a bit confused because I thought the qdot_u_map was needed to properly compute M in the dynamics equation: I forgot that M multiplies u', not u (I was thinking that q' needed to be replaced by u to have things in the proper form)

With that clarified,  I updated the PR to set qdot_u_map to empty dictionary if kinematics are implicit.
Setting `qdot_u_map={}` works since an empty dictionary is a valid thing to substitute. So the rest of the code doesn't have to change. Tests seem to be passing (I made all the test_kane.py run the exact tests  twice, once with explicit kinematics, and once with implicit)

I made it such that a call to `kindiffdict` will now do the explicit solve with a warning if implicit was requested.
`rhs` will also do an LUSolve for the kinematic equations: no warnings there since rhs() documentation does already mention the fact that this will result in explicit equations.

I've noticed there were more test functions and added that case to them. There are a couple of tests that are giving me a headache that I've skipped. They seem to be related to the use of aux and/or velocity constrains (other tests doing that don't fail, just these two, see #TODO in `test_kane2.py`). I'll have to dig more to understand why this is the case. My guess is that it has something to do with the fact that aux speeds create a 2nd instance ? Thoughts ?

**Comment by zouhairm** [comment]
@moorepants here is a rather simple but common formulation of kinematics in aerospace which shows the explicit kinematics struggling with the equations

https://github.com/zouhairm/sympy/blob/example/implicit%20vs%20explicit%20kinematics.ipynb

**Comment by zouhairm** [comment]
@moorepants 

After digging into the failing unit test `test_kane2.py:test_sub_qdot`, I figured out that there is unfortunately still a need for `qdot_u_map`. When you said that:

> To eliminate q' from equation 2), you have to solve 1) for q'. But you can probably set things up like:
> 2 . dynamic: M(q, t)*u'(t) + f(u, q', q, t) = 0

That form of dynamics follows from [equation (13) of your paper](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.1023.2764&rep=rep1&type=pdf), where inertial forces are computed from derivative of velocities which are expressed in the form of `∑ v_i * u_i`  (with `v_i` being the velocity partial and `u_i` the generalized speed).
The way this is mechanized in [_form_fr_star:get_partial_velocity](https://github.com/sympy/sympy/blob/5ea305dd9565ef79743840dba5368c324bb7932d/sympy/physics/mechanics/kane.py#L370) uses  `v = [msubs(vel, self._qdot_u_map) for vel in vlist])` to make sure that the velocities are indeed all expressed as a function of `u_i` and that no `q'_i` appears.

If velocities are left to have `q'` terms in them, then equation (13) needs to be modified to now include `q''` terms as well. Some back of the envelope derivation indicates that one could indeed go through that by computing velocity partials w.r.t. `q'_i` and eventually obtain the inertial force in the form:

*  dynamic: M(q, t)*u'(t) + W(q, t)*q''(t) + f(u, q', q, t) = 0

That would have to then be combined with implicit equations of `q''` (obtained from differentiating the kinematic equations?) and end up with a system of equations of roughly this form:

* kinematic: N(q, t)*q'(t) + g(u, q, t) = 0
* implicit dynamics: 
   - M(q, t)*u'(t) + W(q, t)*q''(t) + f(u, q', q, t) = 0
   - N(q,t)*q''(t) + N'(q,t)*q'(t) + d(g(u,q,t))/dt = 0

That last equation would require a much more careful derivation, and it is unclear if things would be linear in `q''` and `u'` to have a useful matrix factorization to obtain a full system mass matrix and equations of the form`M(q,u,t) [q', q'', u'] = f(q,u,t)`

Unfortunately going through that is beyond the scope of what I was trying to do and I need to actually turn my attention to using the outputs of this (it could be a neat paper though :p). Or maybe I'm overcomplicating this and there is a simple way to treat this case?

---

Fortunately, this is only required if velocities and acceleration of center of masses include `q'` terms.
What I'd like to do to hopefully bring this PR to a close is still move forward with not computing `qdot_u_map` if `explicit_kinematics=False`, but to throw an error if when computing velocity partials `q'` terms are encountered. I've updated the PR with this change. For my use case this is acceptable as I can make sure that velocities and accelerations are only a function of generalized speeds, and hopefully it sets up the basis for future work to derive equations as a function of `q''`

P.S. Looks like there is one issue left to deal with and that is speed constraints where there too there is a jacobian being taken w.r.t. `u` which also requires substituting `q'` - let me know if you have suggestions how to handle that?

**Comment by moorepants** [comment]
@zouhairm I haven't had time to look more. Hope to soon. Thanks!

**Comment by zouhairm** [comment]
@moorepants - I've updated the code to handle the case where dependent or auxiliary velocities are expressed as a function of qdots.
Basically `explicit_kinematics=False` will catch those cases and error out (as opposed to computing the wrong values).

I've updated all tests (except for the slow bicycle one) to run with both True and False for explicit_kinematics, and added logic to expect errors for those cases, but also to 'manually' sub u's instead of qdots to still test the solution. They are all passing now (except for the 'authors' one which I imagine is more procedural)

I've been testing with this code for development purposes for the problem I'm working on and it has been working well. Setting the flag to False basically allows me to make sure that explicit symbolic solutions to kinematics which might be complicated do not get substituted in. It does make it more tedious as one has to properly ensure that velocities and accelerations of points/frames of interests in terms of u's only, but it's worth the upfront work

Please let me know if you have any comments on this that need addressing to merge this PR.

**Comment by zouhairm** [comment]
That makes sense to me and that's what I've been able to glean from the code that indeed the current implementation relies on the substitution of q' by u before differentiation.

What I've done basically is ensure that anywhere substitution of q' (in the code this is `msubs(., self._qdot_u_map)`) is done before taking derivatives (either through partial_velocity or jacobians to compute constraint equations), if any q' is detected an error is raised (`_assert_no_qdot`) to inform the user that 'explicit kinematics' are needed for those problems (i.e. q' as functions of u).

The alternative would be to take the derivatives properly through the chain rule by recognizing that q' are functions of u as you've described. One caveat with doing that I believe is that one ends up with q' or q'' terms in the dynamic equations, which means that kinematics and dynamics are no longer decoupled (related to the comments I made [above](https://github.com/sympy/sympy/issues/22626#issuecomment-1019418067)) 
I think given that complexity it'd probably make sense to leave such an implementation for future PR...

Regarding unit tests: I do think that the current ones do catch some of these subtleties. When I first just ignored the substitution thinking it wasn't needed, a few of the tests were failing as velocity partials were simply wrong. I addressed those first with asserting no qdot in partial_velocity, but then more were still failing due to the velocity constraints requiring jacobians w.r.t u ( `test_sub_qdot` , `test_sub_qdot2` and `test_replace_qdots_in_force` specifically test that this is done properly as those are 'features' that were added in PR #16700 and PR #8948). So I'd say that unit tests are addressing some of these edge cases already.
